objects/src/accounts/delta/storage.rs

```diff
@@ -155,7 +179,98 @@ impl Deserializable for AccountStorageDelta {
             updated_items.push((idx, value));
         }
 
-        Ok(Self { cleared_items, updated_items })
+        // deserialize and validate storage map deltas
+        let num_updated_maps = source.read_u8()? as usize;
+        let mut updated_maps = Vec::with_capacity(num_updated_maps);
+        for _ in 0..num_updated_maps {
+            let idx = source.read_u8()?;
+            let value = StorageMapDelta::read_from(source)?;
+            updated_maps.push((idx, value));
+        }
+
+        Ok(Self {
+            cleared_items,
+            updated_items,
+            updated_maps,
+        })
+    }
+}
+
+// STORAGE MAP DELTA
+// ================================================================================================
+
+/// [StorageMapDelta] stores the differences between two states of account storage maps.
+///
+/// The differences are represented as follows:
+/// - leave updates: represented by `cleared_leaves` and `updated_leaves` field.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct StorageMapDelta {
+    pub cleared_leaves: Vec<Word>,
+    pub updated_leaves: Vec<(Word, Word)>,
+}
+
+impl StorageMapDelta {
+    /// Creates a new storage map delta from the provided cleared and updated leaves.
+    pub fn from(cleared_leaves: Vec<Word>, updated_leaves: Vec<(Word, Word)>) -> Self {
+        StorageMapDelta { cleared_leaves, updated_leaves }
+    }
+
+    /// Checks whether this storage map delta is valid.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any of the cleared or updated leaves is referenced more than once (e.g., updated twice).
+    pub fn validate(&self) -> Result<(), AccountDeltaError> {
+        let mut updated_keys = [];
+
+        updated_keys.sort();
+        for key in updated_keys.windows(2) {
+            if key[0] == key[1] {
+                return Err(AccountDeltaError::DuplicateStorageMapLeaf { key: key[0] });
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns true if storage map delta contains no updates.
+    pub fn is_empty(&self) -> bool {
+        self.cleared_leaves.is_empty() && self.updated_leaves.is_empty()
+    }
+}
+
+impl Serializable for StorageMapDelta {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u8(self.cleared_leaves.len() as u8);
+        for leaf in self.cleared_leaves.iter() {
+            leaf.write_into(target);
+        }
+
+        target.write_u8(self.updated_leaves.len() as u8);
+        for (key, value) in self.updated_leaves.iter() {
+            key.write_into(target);
+            value.write_into(target);
+        }
+    }
```

Comment:
**Request**: Winterfell now has a generic implementation for [`Vec<T>`](https://docs.rs/winter-utils/latest/winter_utils/trait.Serializable.html#impl-Serializable-for-Vec%3CT%3E), this can be just:

```suggestion
    fn write_into<W: ByteWriter>(&self, target: &mut W) {
        self.cleared_leaves.write_into(target);
        self.updated_leaves.write_into(target);
    }
```

Reply:
that's indeed nice